### PR TITLE
Store rate limit values and allow checking if limit was reached

### PR DESCRIPTION
**Pull request by mjanda**
If I'm not mistaken, there isn't a way to get http response directly to check rate limiting headers, so this will store them after every request, allow checking them before doing next request and back-off a little if necessary.

I'm golang novice, so maybe the code is a little bit hairy.

### CODE REVIEW

File: ratelimit.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+type RateLimit struct {
+	LastRequestTime time.Time
+	LimitShort      int
+	LimitLong       int
+	UsageShort      int
+	UsageLong       int
+}
+
+// returns true if rate limit was reached during last X seconds
+func RateLimitReachedDuringLast(seconds int64) bool {
```
Comment (paulmach) [inline ratelimit.go]: How about renaming `RateLimitLast` to `RateLimiting` then define this method on the rate limit struct and call it `Exceeded`.  Then you could do something like `if !strava.RateLimiting.Exceeded()`. 

I think this method should use the information in the RateLimit struct along with the current time to figure out if the rate is exceeded, something like:

```
now := time.Now().Unix()
req := LastRequestTime.Unix()

// same 15 minutes block
if now / 900 == req /900 && UsageShort > LimitShort {
  return false
} 

// same day
if now / 86400 == req / 86400 && UsageLong > LimitLong {
  return false
}

return true

**Comment by paulmach** [comment]
Thanks for submitting this, totally forgot about ratelimiting.

**Comment by mjanda** [comment]
Ok, I went over it and did some changes.
Parsing Date header from response, determining remaining time on server to reset limit and adding this duration to local time (can't be sure local/server times are in sync).
Added function to check how many % of rate limit was reached, so it's possible to put low priority request back into queue.

File: ratelimit.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type RateLimit struct {
+	NextRequestTime time.Time
```
Comment (paulmach) [inline ratelimit.go]: I really feel the `NextRequestTime` stuff is specific to your use case. I think the library should just report the info and let the user figure out how to interpret it.

File: ratelimit.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+func (ratelimit *RateLimit) Exceeded() bool {
+	if time.Now().After(ratelimit.NextRequestTime) {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (ratelimit *RateLimit) FractionReached() float32 {
```
Comment (paulmach) [inline ratelimit.go]: float32 or float64, the eternal question :)  The rest of the library uses float64 everywhere else.

**Comment by paulmach** [comment]
Hi,

I committed this code in ce0cacfd9506c48d6e0510aeb693a8775ba2b353  and then editing the rate limiting to only include Exceeded() and FractionReached().  The use case of NextRequestTime is specialized and encourages making requests at the start of the rate limiting window.  While there is nothing to stop people from doing that, I do not think it is not something the library should directly support or encourage. 

Anyway, thanks for contributing and I'm glad you're making use of the library.

**Pull request closed without merging**